File: app/src/main/java/com/urban/mvvmshowcase/view/activity/AbstractViewModelActivity.java
package com.urban.mvvmshowcase.view.activity;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.urban.mvvmshowcase.viewmodel.ViewModel;

public abstract class AbstractViewModelActivity<T extends ViewModel> extends AppCompatActivity {
    private T mViewModel;

    protected abstract T createViewModel();

    protected T vm() {
        return mViewModel;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        configureViewModel();
    }

    private void configureViewModel() {
        mViewModel = (T) getLastCustomNonConfigurationInstance();
        if (mViewModel == null) {
            mViewModel = createViewModel();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        mViewModel.onShow();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mViewModel.onHide();
    }

    @Override
    public Object onRetainCustomNonConfigurationInstance() {
        return mViewModel;
    }
}

File: app/src/main/java/com/urban/mvvmshowcase/android/base/AbstractViewModelActivity.kt
package com.urban.mvvmshowcase.android.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

abstract class AbstractViewModelActivity<T : Any> : AppCompatActivity() {
    private lateinit var viewModel: T

    protected abstract fun createViewModel(): T

    protected fun vm(): T = viewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = lastCustomNonConfigurationInstance as? T? ?: createViewModel()
    }

    override fun onRetainCustomNonConfigurationInstance() = viewModel
}